web/src/components/storage/VolumeDialog.jsx

```diff
@@ -0,0 +1,743 @@
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, contact SUSE LLC.
+ *
+ * To contact SUSE LLC about this file by physical or electronic mail, you may
+ * find current contact information at www.suse.com.
+ */
+
+// @ts-check
+
+import React, { useReducer } from "react";
+import { Button, Form } from "@patternfly/react-core";
+import { sprintf } from "sprintf-js";
+
+import { _ } from "~/i18n";
+import { compact, useDebounce } from "~/utils";
+import {
+  DEFAULT_SIZE_UNIT, SIZE_METHODS, parseToBytes, splitSize
+} from '~/components/storage/utils';
+import { FsField, MountPathField, SizeOptionsField } from "~/components/storage/VolumeFields";
+import { Popup } from '~/components/core';
+
+/**
+ * @typedef {import ("~/client/storage").Volume} Volume
+ * @typedef {import("~/components/storage/utils").SizeMethod} SizeMethod
+ *
+ * @typedef {object} VolumeFormState
+ * @property {Volume} volume
+ * @property {VolumeFormData} formData
+ * @property {VolumeFormErrors} errors
+ *
+ * @typedef {object} VolumeFormData
+ * @property {number|string} [size]
+ * @property {string} [sizeUnit]
+ * @property {number|string} [minSize]
+ * @property {string} [minSizeUnit]
+ * @property {number|string} [maxSize]
+ * @property {string} [maxSizeUnit]
+ * @property {SizeMethod} sizeMethod
+ * @property {string} mountPath
+ * @property {string} fsType
+ * @property {boolean} snapshots
+ *
+ * @typedef {object} VolumeFormErrors
+ * @property {string|null} missingMountPath
+ * @property {string|null} invalidMountPath
+ * @property {React.ReactElement|null} existingVolume
+ * @property {React.ReactElement|null} existingTemplate
+ * @property {string|null} missingSize
+ * @property {string|null} missingMinSize
+ * @property {string|null} invalidMaxSize
+ *
+ * @typedef {object} VolumeDialogProps
+ * @property {Volume} volume
+ * @property {Volume[]} volumes
+ * @property {Volume[]} templates
+ * @property {boolean} [isOpen=false]
+ * @property {() => void} onCancel
+ * @property {(volume: Volume) => void} onAccept
+ */
+
+/**
+ * Renders the title for the dialog.
+ * @function
+ *
+ * @param {Volume} volume
+ * @param {Volume[]} volumes
+ * @returns {string}
+ */
+const renderTitle = (volume, volumes) => {
+  const isNewVolume = !volumes.includes(volume);
+  const isProductDefined = volume.outline.productDefined;
+  const mountPath = volume.mountPath === "/" ? "root" : volume.mountPath;
+
+  if (isNewVolume && isProductDefined) return sprintf(_("Add %s file system"), mountPath);
+  if (!isNewVolume && isProductDefined) return sprintf(_("Edit %s file system"), mountPath);
+
+  return isNewVolume ? _("Add file system") : _("Edit file system");
+};
+
+class MissingMountPathError {
```

Comment:
I'd expect that the `*Error` names are user for some exceptions. In this case the names should be rather `*Check` as the class does not represent an error but rather a check for error which might or might not happen in the end.

So rather use `MissingMountPathCheck`, `InvalidMountPathCheck`, ...

Reply:
I am not sure if `*Check` is a good name. Those classes are helpers for both to check and to render the error. Maybe we could use something like `*Manager`, that is, `MissingMountPathManager`:

~~~
const errorManager = new MissingMountPathManager("");
errorManager.check() ? errorManager.render(); 
~~~